### PR TITLE
fix: introduce LevelEnv for regions

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by mlutze** [comment]
@magnus-madsen this is an implementation of the idea presented in #6142. It seems to work. I've enabled a a check in Typer that previously caused a lot of issues, and things seem to work properly here

The idea is that we add an `lenv` to unification, which holds the set of regions symbols that we are under. During unification, when we see a region symbol, we check if we are under that region. If so, we leave the symbol as-is. If not, we replace it with `Pure`.

The failing tests are due to not reporting the "right" error. Enabling the effect check means that generalization errors are no longer possible.

This is only a draft; I've left lots of things to clean up. But I'd like an opinion on it before I go through the effort to make it nice.

**Comment by magnus-madsen** [comment]
I think we should look at it on the whiteboard with Jaco and JLS.

**Comment by mlutze** [comment]
So... blocked until August? :'(

**Comment by magnus-madsen** [comment]
No. Let's look at it next week.

**Comment by magnus-madsen** [comment]
Why are generalization errors no longer possible? At least if not related to regions they should be possible, no?

What happens to the Regions phase? Will we remove it and then add levels?

**Comment by magnus-madsen** [comment]
The changes don't seem too bad; except we have to carry around penv/lenv/senv everywhere. 

I suppose the cost of purifyType (purifyVar?) should be cheap.

**Comment by mlutze** [comment]
> Why are generalization errors no longer possible? At least if not related to regions they should be possible, no?

This change adds/activates a check at the top expression of defs to ensure the type and effect match the annotation. That means that anything that would cause a generalization error would be caught at this check instead. Note that generalization errors that are not related to effects are already impossible, since we already have this check for types.


> What happens to the Regions phase? Will we remove it and then add levels?

I think so. But it looks like the `noEscape` check in the Typer is covering all the cases in tests; we need to add some that are actually caught by `Regions`.

**Comment by magnus-madsen** [comment]
Any performance implications?

**Comment by mlutze** [comment]
Round 1: Benchmarking feature-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          4.8 ms (00.1%)
  Tailrec                         13.4 ms (00.1%)
  Reader                          17.8 ms (00.2%)
  VarNumbering                    22.4 ms (00.2%)
  Regions                         28.3 ms (00.3%)
  Safety                          33.6 ms (00.4%)
  Eraser                          38.9 ms (00.4%)
  Simplifier                      52.4 ms (00.6%)
  Stratifier                      65.1 ms (00.7%)
  PatternExhaustiveness           67.8 ms (00.7%)
  EarlyTreeShaker                 72.0 ms (00.8%)
  Reducer                         78.1 ms (00.8%)
  LambdaLift                      90.6 ms (01.0%)
  LateTreeShaker                 103.7 ms (01.1%)
  Kinder                         147.1 ms (01.6%)
  ClosureConv                    162.7 ms (01.7%)
  Lowering                       167.3 ms (01.8%)
  MonomorphEnums                 182.4 ms (01.9%)
  Namer                          193.2 ms (02.0%)
  Redundancy                     243.3 ms (02.6%)
  Instances                      256.7 ms (02.7%)
  Weeder                         318.8 ms (03.4%)
  Optimizer                      441.7 ms (04.7%)
  Resolver                       571.1 ms (06.0%)
  Parser                         621.3 ms (06.6%)
  Monomorph                      867.2 ms (09.1%)
  JvmBackend                     1494.4 ms (15.8%)
  Typer                          3126.1 ms (33.0%)

Finished 10 iterations on 64,934 lines of code in 79 seconds.
Round 1: Benchmarking master-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          4.1 ms (00.0%)
  Reader                          16.8 ms (00.2%)
  VarNumbering                    20.4 ms (00.2%)
  Regions                         23.7 ms (00.3%)
  Eraser                          27.7 ms (00.3%)
  Tailrec                         30.5 ms (00.3%)
  Safety                          34.2 ms (00.4%)
  Simplifier                      49.1 ms (00.5%)
  Stratifier                      53.9 ms (00.6%)
  EarlyTreeShaker                 64.8 ms (00.7%)
  PatternExhaustiveness           65.3 ms (00.7%)
  LambdaLift                      85.5 ms (00.9%)
  Reducer                         89.2 ms (00.9%)
  ClosureConv                     93.8 ms (01.0%)
  LateTreeShaker                  98.5 ms (01.0%)
  Kinder                         152.8 ms (01.6%)
  Lowering                       162.1 ms (01.7%)
  MonomorphEnums                 178.0 ms (01.9%)
  Namer                          182.6 ms (01.9%)
  Redundancy                     241.0 ms (02.5%)
  Weeder                         257.6 ms (02.7%)
  Instances                      261.1 ms (02.8%)
  Optimizer                      415.2 ms (04.4%)
  Resolver                       504.6 ms (05.3%)
  Parser                         700.1 ms (07.4%)
  Monomorph                      918.7 ms (09.7%)
  JvmBackend                     1276.3 ms (13.5%)
  Typer                          3450.7 ms (36.5%)

Finished 10 iterations on 64,934 lines of code in 81 seconds.
Round 2: Benchmarking feature-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          5.0 ms (00.0%)
  Tailrec                         13.6 ms (00.1%)
  Reader                          23.6 ms (00.2%)
  VarNumbering                    25.9 ms (00.2%)
  Regions                         27.0 ms (00.2%)
  Eraser                          41.6 ms (00.4%)
  Safety                          48.9 ms (00.4%)
  LambdaLift                      50.0 ms (00.5%)
  Simplifier                      58.0 ms (00.5%)
  PatternExhaustiveness           72.3 ms (00.7%)
  Stratifier                      76.2 ms (00.7%)
  Reducer                         84.4 ms (00.8%)
  EarlyTreeShaker                 85.5 ms (00.8%)
  ClosureConv                    110.3 ms (01.0%)
  LateTreeShaker                 117.7 ms (01.1%)
  Kinder                         175.4 ms (01.6%)
  Lowering                       195.6 ms (01.8%)
  Namer                          228.6 ms (02.1%)
  MonomorphEnums                 236.3 ms (02.1%)
  Instances                      293.8 ms (02.7%)
  Redundancy                     294.6 ms (02.7%)
  Weeder                         352.5 ms (03.2%)
  Optimizer                      490.9 ms (04.4%)
  Resolver                       561.7 ms (05.1%)
  Parser                         836.3 ms (07.6%)
  Monomorph                      1055.0 ms (09.5%)
  JvmBackend                     1805.1 ms (16.3%)
  Typer                          3709.2 ms (33.5%)

Finished 10 iterations on 64,934 lines of code in 92 seconds.
Round 2: Benchmarking master-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          4.4 ms (00.0%)
  Tailrec                         13.3 ms (00.1%)
  Reader                          22.4 ms (00.2%)
  Regions                         24.5 ms (00.2%)
  VarNumbering                    24.5 ms (00.2%)
  Eraser                          31.4 ms (00.3%)
  Safety                          33.6 ms (00.3%)
  LambdaLift                      53.0 ms (00.5%)
  Simplifier                      55.9 ms (00.5%)
  PatternExhaustiveness           70.2 ms (00.6%)
  Stratifier                      73.5 ms (00.7%)
  EarlyTreeShaker                 80.5 ms (00.7%)
  Reducer                         87.8 ms (00.8%)
  ClosureConv                    137.0 ms (01.2%)
  LateTreeShaker                 150.0 ms (01.3%)
  Kinder                         151.0 ms (01.4%)
  Lowering                       175.5 ms (01.6%)
  MonomorphEnums                 218.5 ms (02.0%)
  Namer                          241.4 ms (02.2%)
  Redundancy                     301.0 ms (02.7%)
  Instances                      316.1 ms (02.8%)
  Weeder                         328.2 ms (02.9%)
  Optimizer                      496.1 ms (04.5%)
  Resolver                       604.8 ms (05.4%)
  Parser                         917.4 ms (08.2%)
  Monomorph                      982.9 ms (08.8%)
  JvmBackend                     1498.0 ms (13.4%)
  Typer                          4047.2 ms (36.3%)

Finished 10 iterations on 64,934 lines of code in 96 seconds.
Round 3: Benchmarking feature-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          5.4 ms (00.0%)
  Tailrec                         13.8 ms (00.1%)
  VarNumbering                    24.9 ms (00.2%)
  Eraser                          32.5 ms (00.3%)
  Reader                          38.4 ms (00.3%)
  Safety                          38.6 ms (00.3%)
  Regions                         52.3 ms (00.5%)
  Simplifier                      56.3 ms (00.5%)
  Stratifier                      65.5 ms (00.6%)
  EarlyTreeShaker                 71.8 ms (00.6%)
  PatternExhaustiveness           75.6 ms (00.7%)
  Reducer                         79.0 ms (00.7%)
  LambdaLift                      85.4 ms (00.8%)
  LateTreeShaker                 114.7 ms (01.0%)
  ClosureConv                    156.7 ms (01.4%)
  Kinder                         175.2 ms (01.5%)
  Lowering                       187.9 ms (01.7%)
  MonomorphEnums                 192.4 ms (01.7%)
  Namer                          245.8 ms (02.2%)
  Redundancy                     273.7 ms (02.4%)
  Instances                      340.5 ms (03.0%)
  Weeder                         381.9 ms (03.4%)
  Optimizer                      456.4 ms (04.0%)
  Resolver                       585.3 ms (05.2%)
  Parser                         974.8 ms (08.6%)
  Monomorph                      995.1 ms (08.8%)
  JvmBackend                     1453.0 ms (12.8%)
  Typer                          4189.8 ms (36.9%)

Finished 10 iterations on 64,934 lines of code in 99 seconds.
Round 3: Benchmarking master-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          4.5 ms (00.0%)
  Tailrec                         14.0 ms (00.1%)
  Reader                          21.1 ms (00.2%)
  VarNumbering                    27.9 ms (00.3%)
  Regions                         28.2 ms (00.3%)
  Eraser                          35.8 ms (00.3%)
  Safety                          48.7 ms (00.4%)
  Simplifier                      57.3 ms (00.5%)
  PatternExhaustiveness           64.9 ms (00.6%)
  EarlyTreeShaker                 79.2 ms (00.7%)
  Stratifier                      83.5 ms (00.8%)
  LambdaLift                      87.5 ms (00.8%)
  ClosureConv                     93.8 ms (00.9%)
  LateTreeShaker                 121.7 ms (01.1%)
  Reducer                        142.2 ms (01.3%)
  Kinder                         156.0 ms (01.4%)
  Lowering                       177.7 ms (01.6%)
  MonomorphEnums                 185.4 ms (01.7%)
  Namer                          220.0 ms (02.0%)
  Redundancy                     276.6 ms (02.5%)
  Instances                      297.1 ms (02.7%)
  Weeder                         337.2 ms (03.1%)
  Optimizer                      539.9 ms (05.0%)
  Resolver                       552.8 ms (05.1%)
  Parser                         872.3 ms (08.0%)
  Monomorph                      1030.6 ms (09.5%)
  JvmBackend                     1652.9 ms (15.2%)
  Typer                          3680.9 ms (33.8%)

Finished 10 iterations on 64,934 lines of code in 92 seconds.
Round 4: Benchmarking feature-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          5.1 ms (00.0%)
  Tailrec                         13.7 ms (00.1%)
  VarNumbering                    22.0 ms (00.2%)
  Reader                          22.2 ms (00.2%)
  Regions                         27.1 ms (00.2%)
  Safety                          44.2 ms (00.4%)
  LambdaLift                      50.9 ms (00.5%)
  Simplifier                      56.3 ms (00.5%)
  EarlyTreeShaker                 73.5 ms (00.7%)
  Stratifier                      82.6 ms (00.7%)
  Reducer                         82.8 ms (00.8%)
  Eraser                          82.9 ms (00.8%)
  PatternExhaustiveness           84.8 ms (00.8%)
  ClosureConv                    107.0 ms (01.0%)
  LateTreeShaker                 116.6 ms (01.1%)
  Lowering                       183.0 ms (01.7%)
  MonomorphEnums                 185.5 ms (01.7%)
  Kinder                         203.1 ms (01.8%)
  Namer                          224.5 ms (02.0%)
  Redundancy                     266.8 ms (02.4%)
  Weeder                         333.5 ms (03.0%)
  Instances                      346.3 ms (03.1%)
  Optimizer                      474.6 ms (04.3%)
  Resolver                       644.0 ms (05.8%)
  Parser                         776.5 ms (07.0%)
  Monomorph                      988.1 ms (09.0%)
  JvmBackend                     1710.8 ms (15.5%)
  Typer                          3822.9 ms (34.7%)

Finished 10 iterations on 64,934 lines of code in 93 seconds.
Round 4: Benchmarking master-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.1 ms (00.0%)
  Deriver                          4.6 ms (00.0%)
  Tailrec                         12.9 ms (00.1%)
  Reader                          20.4 ms (00.2%)
  VarNumbering                    24.4 ms (00.2%)
  Regions                         27.5 ms (00.2%)
  Eraser                          32.3 ms (00.3%)
  Safety                          48.6 ms (00.4%)
  LambdaLift                      52.7 ms (00.5%)
  Stratifier                      67.0 ms (00.6%)
  EarlyTreeShaker                 69.9 ms (00.6%)
  PatternExhaustiveness           70.0 ms (00.6%)
  Simplifier                      86.0 ms (00.8%)
  Reducer                        105.8 ms (01.0%)
  ClosureConv                    121.1 ms (01.1%)
  LateTreeShaker                 136.7 ms (01.2%)
  MonomorphEnums                 176.5 ms (01.6%)
  Lowering                       188.2 ms (01.7%)
  Kinder                         221.2 ms (02.0%)
  Redundancy                     235.3 ms (02.1%)
  Namer                          244.2 ms (02.2%)
  Instances                      273.2 ms (02.5%)
  Weeder                         360.7 ms (03.3%)
  Optimizer                      492.5 ms (04.4%)
  Resolver                       556.3 ms (05.0%)
  Parser                         837.5 ms (07.6%)
  Monomorph                      972.4 ms (08.8%)
  JvmBackend                     1632.8 ms (14.7%)
  Typer                          4002.9 ms (36.1%)

Finished 10 iterations on 64,934 lines of code in 94 seconds.
Round 5: Benchmarking feature-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          4.2 ms (00.0%)
  Tailrec                         13.6 ms (00.1%)
  Reader                          21.2 ms (00.2%)
  VarNumbering                    24.8 ms (00.2%)
  Regions                         27.1 ms (00.2%)
  Eraser                          35.5 ms (00.3%)
  Safety                          44.8 ms (00.4%)
  LambdaLift                      53.3 ms (00.5%)
  Simplifier                      55.9 ms (00.5%)
  PatternExhaustiveness           68.0 ms (00.6%)
  Stratifier                      68.8 ms (00.6%)
  EarlyTreeShaker                 78.7 ms (00.7%)
  ClosureConv                    104.0 ms (00.9%)
  Reducer                        106.0 ms (00.9%)
  LateTreeShaker                 113.6 ms (01.0%)
  MonomorphEnums                 189.6 ms (01.7%)
  Lowering                       189.6 ms (01.7%)
  Kinder                         196.0 ms (01.8%)
  Namer                          220.2 ms (02.0%)
  Redundancy                     254.0 ms (02.3%)
  Instances                      270.7 ms (02.4%)
  Weeder                         318.2 ms (02.8%)
  Optimizer                      455.7 ms (04.1%)
  Resolver                       587.9 ms (05.3%)
  Parser                         968.6 ms (08.7%)
  Monomorph                      989.0 ms (08.8%)
  JvmBackend                     1742.4 ms (15.6%)
  Typer                          3994.1 ms (35.7%)

Finished 10 iterations on 64,934 lines of code in 94 seconds.
Round 5: Benchmarking master-flix.jar
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
====================== Flix Compiler Phases ======================

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                         13.5 ms (00.1%)
  Tailrec                         14.2 ms (00.1%)
  Reader                          21.8 ms (00.2%)
  Regions                         26.4 ms (00.2%)
  Safety                          34.7 ms (00.3%)
  VarNumbering                    42.1 ms (00.4%)
  LambdaLift                      56.2 ms (00.5%)
  Simplifier                      62.4 ms (00.6%)
  PatternExhaustiveness           63.7 ms (00.6%)
  Stratifier                      78.0 ms (00.7%)
  EarlyTreeShaker                 78.3 ms (00.7%)
  Reducer                         83.3 ms (00.8%)
  Eraser                          88.3 ms (00.8%)
  ClosureConv                    110.2 ms (01.0%)
  LateTreeShaker                 112.2 ms (01.0%)
  Lowering                       173.0 ms (01.6%)
  MonomorphEnums                 191.1 ms (01.7%)
  Kinder                         210.7 ms (01.9%)
  Namer                          241.8 ms (02.2%)
  Redundancy                     273.3 ms (02.5%)
  Instances                      276.3 ms (02.5%)
  Weeder                         350.8 ms (03.2%)
  Optimizer                      525.9 ms (04.7%)
  Resolver                       642.0 ms (05.8%)
  Parser                         843.0 ms (07.6%)
  Monomorph                      1095.7 ms (09.9%)
  JvmBackend                     1654.7 ms (14.9%)
  Typer                          3733.6 ms (33.6%)

Finished 10 iterations on 64,934 lines of code in 94 seconds.

**Comment by mlutze** [comment]
Seems inconclusive

File: main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala [hunk 1]
```diff
@@ -31,7 +31,9 @@ object EffUnification {
   /**
     * Returns the most general unifier of the two given Boolean formulas `tpe1` and `tpe2`.
     */
-  def unify(tpe1: Type, tpe2: Type, renv0: RigidityEnv)(implicit flix: Flix): Result[(Substitution, List[Ast.BroadEqualityConstraint]), UnificationError] = {
+  def unify(tpe10: Type, tpe20: Type, renv0: RigidityEnv, lenv: LevelEnv)(implicit flix: Flix): Result[(Substitution, List[Ast.BroadEqualityConstraint]), UnificationError] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:34]: Why these names tpe10? tpe20?

File: main/src/ca/uwaterloo/flix/language/phase/unification/EqualityEnvironment.scala [hunk 2]
```diff
@@ -85,7 +85,7 @@ object EqualityEnvironment {
     val insts = eqEnv(cst.sym)
     insts.iterator.flatMap { // TODO ASSOC-TYPES generalize this pattern (also in monomorph)
       inst =>
-        Unification.unifyTypes(arg, inst.arg, renv).toOption.map {
+        Unification.unifyTypes(arg, inst.arg, renv, LevelEnv.Top).toOption.map { // TODO level env?
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/EqualityEnvironment.scala:88]: TODO?

File: main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala [hunk 3]
```diff
@@ -1154,7 +1154,7 @@ class TestTyper extends AnyFunSuite with TestUtils {
         |
       """.stripMargin
     val result = compile(input, Options.TestWithLibMin)
-    expectError[TypeError.ImpureDeclaredAsPure](result)
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala:1157]: Do we need to delete ImpureDeclaredAsPure? If it is no longer used?

**Comment by magnus-madsen** [comment]
What is also the situation with the error messages? i.e. impure declared as pure. Why does that fail?

File: main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala (hunk 1 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:34]: Using the convention that initial values have the `0` suffix. I know you hate it :)

Would you prefer `tpe1_0`?

Or some other idea?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:34]: Why can it not just be tpe1 and tpe2? :)

File: main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala (hunk 3 above)
Comment (mlutze) [reply main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala:1157]: I was thinking that we could detect if there's an error at the top level and still use this error in that case.

Comment (mlutze) [reply main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala:1157]: (in a future PR)

**Comment by mlutze** [comment]
> What is also the situation with the error messages? i.e. impure declared as pure. Why does that fail?

This is due to the change in line 1900 of the Typer. It means that effects are checked against the function signature before we get to the generalization error check, which is what detects that specific error.

**Comment by mlutze** [comment]
> > What is also the situation with the error messages? i.e. impure declared as pure. Why does that fail?
> 
> This is due to the change in line 1900 of the Typer. It means that effects are checked against the function signature before we get to the generalization error check, which is what detects that specific error.

In a future PR we can modify that specific check to issue a more precise error.

File: main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala (hunk 1 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:34]: Because we introduce tpe1 and tpe2 just below.

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**